### PR TITLE
Add component stack info to key validation warnings

### PR DESCRIPTION
**Pull request by keyanzhang**
This PR implements #6790. ~~I rewrote some of the tests from JSX to `createElement` calls since #6398 added source locations at JSX compile-time and hard-coding line numbers in tests doesn't sound good.~~

- [x] "Encountered two children with the same key" error
  - [x] `ReactChildReconciler`
  - [x] `flattenChildren`
- [x] "Each child in an array or iterator should have a unique 'key' prop"
  - [x] `ReactElementValidator`
- [x] Switch tests back to JSX and normalize code location info in tests
  - [x] `ReactElementValidator-test`
  - [x] `ReactChildReconciler-test`
  - [x] `ReactMultiChild-test`
- [x] Clean up `ReactElementValidator`
  - [x] Update `ReactElementValidator-test`
- [x] Fix `ReactChildReconciler`
- [x] Refactor `flattenChildren` and add test cases 
  - [x] Test cases
- [x] Implement `ReactComponentTreeDevtool.getStackAddendumByID`
- [x] Pass `debugID` to `ReactTransitionChildMapping.getChildMapping `
- [ ] Improve `ReactElementValidator` Error Message

CC @spicyj

### CODE REVIEW

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 1]
```diff
@@ -72,10 +72,11 @@ function validateExplicitKey(element, parentType) {
   warning(
     false,
     'Each child in an array or iterator should have a unique "key" prop.' +
-    '%s%s%s',
+    '%s%s%s%s',
     addenda.parentOrOwner || '',
     addenda.childOwner || '',
-    addenda.url || ''
+    addenda.url || '',
```
Comment (spicyj) [inline src/isomorphic/classic/element/ReactElementValidator.js]: We shouldn't need these other addenda now since the stack should cover it. Can you remove the code for them? We still will want to dedupe the warnings based on owner though as we're currently doing.

File: src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js [hunk 2]
```diff
@@ -84,37 +84,81 @@ describe('ReactElementValidator', function() {
     var Anonymous = React.createClass({
       displayName: undefined,
       render: function() {
-        return <div />;
+        return React.createElement('div');
```
Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js]: Instead of changing these to not use JSX, can you just make the tests like I have in other files, replacing the filename and line number with `(**)`?

File: src/renderers/shared/stack/reconciler/__tests__/ReactChildReconciler-test.js [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+
+    ReactTestUtils.renderIntoDocument(React.createElement(GrandParent));
+
+    expect(console.error.argsForCall.length).toBe(1);
+    expect(console.error.argsForCall[0][0]).toBe(
+      'Warning: flattenChildren(...): ' +
+      'Encountered two children with the same key, `1`. ' +
+      'Child keys must be unique; when two children share a key, ' +
+      'only the first child will be used.\n' +
+      '    in Unknown\n' + // cuz we are using `ReactTestUtils.renderIntoDocument`
```
Comment (spicyj) [inline src/renderers/shared/stack/reconciler/__tests__/ReactChildReconciler-test.js]: ReactTestUtils.renderIntoDocument shouldn't affect display names; why does it here?

File: src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js (hunk 2 above)
Comment (keyanzhang) [inline src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js]: Oh yeah, sorry I didn't see your regexp. Sure!

File: src/renderers/shared/stack/reconciler/__tests__/ReactChildReconciler-test.js (hunk 3 above)
Comment (spicyj) [inline src/renderers/shared/stack/reconciler/__tests__/ReactChildReconciler-test.js]: I'm pretty sure this is unrelated to ReactTestUtils.renderIntoDocument; commenting above.

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js [hunk 4]
```diff
@@ -27,8 +28,9 @@ function instantiateChild(childInstances, child, name) {
       keyUnique,
       'flattenChildren(...): Encountered two children with the same key, ' +
       '`%s`. Child keys must be unique; when two children share a key, only ' +
-      'the first child will be used.',
-      KeyEscapeUtils.unescape(name)
+      'the first child will be used.%s',
+      KeyEscapeUtils.unescape(name),
+      ReactComponentTreeDevtool.getCurrentStackAddendum(childInstances[name])
```
Comment (spicyj) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: This is not a React element; it is an internal React component instance. This is why you got the useless "in Unknown" stack frame. I think `childInstances[name]._currentElement` is pretty close to what you want here, but that will actually not give you a useful stack if `childInstances[name]` is a text component. This error message could also be misleading if the duplicate keys are on components of different types since you'd only see one of them.

Instead, I think it probably makes sense to only show the stack to the parent component – but it is actually going to be a bit complicated to thread that information through. getCurrentStackAddendum always looks at the *current* owner (hence its name), but during reconciliation here we shouldn't rely on the owner being set. I'm actually surprised that it even is in your test case. Let's add a new function ReactComponentTreeDevtool.getStackAddendum(id) that takes a debug ID and prints the parents of that instance. We'll want to reuse some of the code from getCurrentStackAddendum but nothing about topElement or ReactCurrentOwner.

You'll probably want to add a parameter to ReactChildReconciler.instantiateChildren that takes the debug ID of the parent component, then pass that from ReactMultiChild.

File: src/renderers/shared/stack/reconciler/__tests__/ReactMultiChild-test.js [hunk 5]
```diff
@@ -148,5 +148,60 @@ describe('ReactMultiChild', function() {
       expect(mockMount.mock.calls.length).toBe(2);
       expect(mockUnmount.mock.calls.length).toBe(1);
     });
+
+    it('should warn for duplicated keys with component stack info', function() {
+      // NOTE: We're explicitly not using JSX here. This is intended to test
```
Comment (spicyj) [inline src/renderers/shared/stack/reconciler/__tests__/ReactMultiChild-test.js]: Again, let's use JSX and do a regex replace to normalize it so we can check equality.

File: src/shared/utils/flattenChildren.js [hunk 6]
```diff
@@ -29,8 +30,9 @@ function flattenSingleChildIntoContext(traverseContext, child, name) {
       keyUnique,
       'flattenChildren(...): Encountered two children with the same key, ' +
       '`%s`. Child keys must be unique; when two children share a key, only ' +
-      'the first child will be used.',
-      KeyEscapeUtils.unescape(name)
+      'the first child will be used.%s',
+      KeyEscapeUtils.unescape(name),
+      ReactComponentTreeDevtool.getCurrentStackAddendum(result[name])
```
Comment (spicyj) [inline src/shared/utils/flattenChildren.js]: Likewise, you'll want to thread through a debug ID for the parent component here. It looks like there are several places that use flattenChildren so let's add a test for each one to make sure that the stack makes it through correctly.

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js (hunk 4 above)
Comment (keyanzhang) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Yeah it makes sense to me. I thought the issue with `Unknown` was about the name of the node but I actually got confused by element vs. instance here. Thanks!

File: src/shared/utils/flattenChildren.js (hunk 6 above)
Comment (keyanzhang) [inline src/shared/utils/flattenChildren.js]: Sure!

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js [hunk 7]
```diff
@@ -50,12 +52,31 @@ var ReactChildReconciler = {
...
     if (nestedChildNodes == null) {
       return null;
     }
     var childInstances = {};
-    traverseAllChildren(nestedChildNodes, instantiateChild, childInstances);
+
+    if (__DEV__) {
+      traverseAllChildren(
+        nestedChildNodes,
+        (childInsts, child, name) => instantiateChild(
```
Comment (keyanzhang) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: I tried to put `selfDebugID` as the first arg of `instantiateChild` and to use `instantiateChild.bind(null, selfDebugID)` here, but that means we need to call `bind` as well in the production build. Do you think the current approach (arrowed currying, if that's a term) is acceptable? @spicyj

**Comment by facebook-github-bot-3** [comment]
@keyanzhang updated the pull request.

**Comment by facebook-github-bot-0** [comment]
@keyanzhang updated the pull request.

File: src/addons/transitions/ReactTransitionChildMapping.js [hunk 8]
```diff
@@ -25,6 +25,9 @@ var ReactTransitionChildMapping = {
     if (!children) {
       return children;
     }
+    // TODO: `flattenChildren` now takes an extra `selfDebugID` argument
+    // for looking up the component stack in dev build.
+    // Need to figure out a way to get the `_debugID` here.
     return flattenChildren(children);
```
Comment (keyanzhang) [inline src/addons/transitions/ReactTransitionChildMapping.js]: @spicyj what's a good way to get and feed `debugID` to `flattenChildren` here? The only place that calls `ReactTransitionChildMapping.getChildMapping` (syntactic sugar for `flattenChildren`) is here: https://github.com/facebook/react/blob/master/src/addons/transitions/ReactTransitionGroup.js#L41. In other words, how do I/can I get a debugID from a component?

File: src/addons/transitions/ReactTransitionChildMapping.js [hunk 9]
```diff
@@ -25,6 +25,9 @@ var ReactTransitionChildMapping = {
     if (!children) {
       return children;
     }
+    // TODO: `flattenChildren` now takes an extra `selfDebugID` argument
```
Comment (keyanzhang) [inline src/addons/transitions/ReactTransitionChildMapping.js]: @spicyj what's a good way to get and feed `debugID` to `flattenChildren` here? The only place that calls `ReactTransitionChildMapping.getChildMapping` (syntactic sugar for `flattenChildren`) is here: https://github.com/facebook/react/blob/master/src/addons/transitions/ReactTransitionGroup.js#L41. In other words, how do I/can I get a debugID from a component?

Comment (keyanzhang) [inline src/addons/transitions/ReactTransitionChildMapping.js]: Also, seems like the only other place we use `flattenChildren` is in `ReactMultiChild` and there's already a test for that in `ReactMultiChild-test`. Interestingly, when `ReactChildReconciler` finds duplicated keys, the [error message](https://github.com/facebook/react/blob/master/src/renderers/shared/stack/reconciler/ReactChildReconciler.js#L28) talks about "flattenChildren".

**Comment by keyanzhang** [comment]
This PR should be ready for review. I refactored `ReactElementValidator` (inlined a function) but didn't change the warning message since the current message format is more explicit than the component stack info. Please feel free to let me know if you want me to change it. Thanks @spicyj!

**Comment by facebook-github-bot-1** [comment]
@keyanzhang updated the pull request.

**Comment by facebook-github-bot-5** [comment]
@keyanzhang updated the pull request.

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js (hunk 7 above)
Comment (spicyj) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Yeah, this seems fine to me. (I think calling it `parentDebugID` might be a little clearer, but this is also oaky.)

File: src/addons/transitions/ReactTransitionChildMapping.js (hunk 9 above)
Comment (spicyj) [inline src/addons/transitions/ReactTransitionChildMapping.js]:
```
var internalInstance = ReactInstanceMap.get(this);
internalInstance._debugID
```

should work. Maybe someday we'll have a public API to get this that anyone could use but for now I think it's okay for us to get the internal instance for this here.

File: src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js [hunk 10]
```diff
@@ -95,9 +99,10 @@ describe('ReactElementValidator', function() {
     ReactTestUtils.renderIntoDocument(<Anonymous>{divs}</Anonymous>);
 
     expect(console.error.argsForCall.length).toBe(1);
-    expect(console.error.argsForCall[0][0]).toBe(
+    expect(normalizeCodeLocInfo(console.error.argsForCall[0][0])).toBe(
       'Warning: Each child in an array or iterator should have a unique ' +
-      '"key" prop. See https://fb.me/react-warning-keys for more information.'
+      '"key" prop. See https://fb.me/react-warning-keys for more information.\n' +
+      '    in div (at **)'
```
Comment (spicyj) [inline src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js]: It would be nice if this had both the parent component (Anonymous, or "Unknown" as we'll know it as in the devtools) and the child (div) here. Is that possible? I guess we would probably have to make getCurrentStackAddendum take both elements as arguments.

File: src/renderers/shared/devtools/ReactComponentTreeDevtool.js [hunk 11]
```diff
@@ -192,11 +192,17 @@ var ReactComponentTreeDevtool = {
 
     var currentOwner = ReactCurrentOwner.current;
     var id = currentOwner && currentOwner._debugID;
+
+    info += ReactComponentTreeDevtool.getStackAddendumByID(id);
```
Comment (spicyj) [inline src/renderers/shared/devtools/ReactComponentTreeDevtool.js]: Nice.

**Comment by spicyj** [comment]
Did you change anything in ReactElementValidator other than adding the stack? It's hard to tell because the code moved around.

**Comment by spicyj** [comment]
(Also, pretty sure I meant to do needs-revision on this last time! Just a few more questions and I think we'll be good.)

File: src/renderers/shared/stack/reconciler/ReactChildReconciler.js (hunk 7 above)
Comment (keyanzhang) [inline src/renderers/shared/stack/reconciler/ReactChildReconciler.js]: Hmm, I thought about this; `self` is more clear to me since we are talking about the current component which happens to be the parent for the children here.

**Comment by facebook-github-bot-7** [comment]
@keyanzhang updated the pull request.

File: src/addons/transitions/ReactTransitionGroup.js [hunk 12]
```diff
@@ -38,6 +39,7 @@ var ReactTransitionGroup = React.createClass({
 
   getInitialState: function() {
     return {
+      // TODO: can we get useful debug information to show at this point?
```
Comment (keyanzhang) [inline src/addons/transitions/ReactTransitionGroup.js]: Interestingly, ReactInstanceMap doesn't have information for the component at this point. That's probably because the component hasn't been mounted.

File: src/addons/transitions/__tests__/ReactTransitionGroup-test.js [hunk 13]
```diff
@@ -269,4 +273,33 @@ describe('ReactTransitionGroup', function() {
...
+    });
+
+    ReactDOM.render(<Component />, container);
+
+    expect(console.error.argsForCall.length).toBe(2);
+    expect(console.error.argsForCall[0][0]).toBe(
+      'Warning: flattenChildren(...): ' +
+      'Encountered two children with the same key, `1`. ' +
+      'Child keys must be unique; when two children share a key, ' +
+      'only the first child will be used.'
```
Comment (keyanzhang) [inline src/addons/transitions/__tests__/ReactTransitionGroup-test.js]: The first warning has no stack info. I think this is because it is triggered by `getInitialState` and we don't have stack info about the component yet. Moving `ReactInstanceMap.set` from mount-time to construction-type might fix this issue, but I don't know if that makes sense. @spicyj

File: src/isomorphic/classic/element/ReactElementValidator.js [hunk 14]
```diff
@@ -64,67 +78,36 @@ function validateExplicitKey(element, parentType) {
...
+    childOwner =
       ` It was passed a child from ${element._owner.getName()}.`;
   }
 
-  return addenda;
+  warning(
+    false,
+    'Each child in an array or iterator should have a unique "key" prop.' +
+    '%s%s See https://fb.me/react-warning-keys for more information.%s',
+    currentComponentErrorInfo,
```
Comment (keyanzhang) [inline src/isomorphic/classic/element/ReactElementValidator.js]: I removed the error info object packing and unpacking logic and moved the `memoizer` to `validateExplicitKey`. Other than that there shouldn't be any (behavioral) changes.

File: src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js (hunk 10 above)
Comment (keyanzhang) [inline src/isomorphic/classic/element/__tests__/ReactElementValidator-test.js]: Yeah, that sounds like an interesting sub-project -- I'd love to tweak and improve the error messages, but would it be better to do it in a separate PR? I guess we can talk about this on Monday.

**Comment by spicyj** [comment]
looks good to me, feel free to "squash and merge"

**Comment by tnrich** [comment]
Yay I am so stoked for this to be merged in! Great work react team!
